app/ts/simulation/priceEstimator.ts

```diff
@@ -21,99 +20,92 @@ export class PriceEstimator {
+					quoteToken,
+					price: 10n ** 18n
 				})
 				continue
 			}
-			const swapTransactions = [
-				{
-					type: '1559' as const,
-					from: sender,
-					chainId: chainId,
-					nonce: transactionCount,
-					maxFeePerGas: 0n,
-					maxPriorityFeePerGas: 0n,
-					gas: 15000000n,
-					to: UNISWAP_V2_ROUTER_ADDRESS,
-					value: 10n ** 18n,
-					input: stringToUint8Array(swapInterface.encodeFunctionData(
-						'swapExactETHForTokens',
-						[amountOutMin, [addressString(CHAINS[chainString].weth), addressString(token.token)], addressString(sender), deadline]
-					)),
-					accessList: [],
-				},
-				{
-					type: '1559' as const,
-					from: sender,
-					chainId: chainId,
-					nonce: transactionCount + 1n,
-					maxFeePerGas: 0n,
-					maxPriorityFeePerGas: 0n,
-					gas: 15000000n,
-					to: token.token,
-					value: 0n,
-					input: stringToUint8Array(swapInterface.encodeFunctionData('approve', [addressString(UNISWAP_V2_ROUTER_ADDRESS), 2n ** 127n] )),
-					accessList: [],
-				},
-				{
-					type: '1559' as const,
-					from: sender,
-					chainId: chainId,
-					nonce: transactionCount + 2n,
-					maxFeePerGas: 0n,
-					maxPriorityFeePerGas: 0n,
-					gas: 15000000n,
-					to: UNISWAP_V2_ROUTER_ADDRESS,
-					value: 0n,
-					input: stringToUint8Array(swapInterface.encodeFunctionData(
-						'swapTokensForExactETH',
-						[10n ** 18n / 2n, 2n ** 127n, [addressString(token.token), addressString(CHAINS[chainString].weth)], addressString(sender), deadline]
-					)),
-					accessList: [],
-				},
-			]
-			const results = await this.ethereum.multicall(swapTransactions, block.number + 1n)
-			if (results.length !== 3) throw ('invalid multicall result')
-			if (results[2].statusCode === 'success') {
-				const parsed = swapInterface.decodeFunctionResult('swapTokensForExactETH', results[2].returnValue)
-				const inOut = parsed.toObject() as { amounts: bigint[] } // TODO, change to funtype
-				if (inOut.amounts.length != 2) return []
-				if(inOut.amounts[0] <= 0n || inOut.amounts[1] <= 0n || token.decimals <= 0n) return []
-				inOutResults.push( {
-					token: addressString(token.token ),
-					inOutAmount: [inOut.amounts[1], inOut.amounts[0]],
-					decimals: token.decimals
-				})
+			const poolAddresses = calculateUniswapPools(token.address, CHAINS[chainString].weth)
+			const uniswapSpotCalls = getUniswapSpotCalls(token.address, quoteToken.address, poolAddresses)
+
+			const callData = stringToUint8Array(IMulticall3.encodeFunctionData('aggregate3', [uniswapSpotCalls]))
+			const callTransaction = {
+				type: '1559',
+				to: MULTICALL3,
+				value: 0n,
+				input: callData,
 			}
+			const multicallReturnData: { success: boolean, returnData: string }[] = IMulticall3.decodeFunctionResult('aggregate3', await this.ethereum.call(callTransaction, 'latest'))[0]
+
+			const prices = multicallReturnData.reduce((prices, { success, returnData }, index): { price: bigint, k: bigint }[] => {
+				if (index > 4) return prices
```

Comment:
Where does its `4` come from?

Reply:
There are 5 pool variants, that have calls + then balanceOf calls to query v3 pool balances, so calls after index 4 (5th and last pool) are balanceOf calls.